drivers/nvdimm/pmem.c

```diff
@@ -45,11 +65,352 @@ static struct nd_region *to_region(struct pmem_device *pmem)
+		    bvec.bv_offset, bvec.bv_len, dir);
+		if (dma_mapping_error(dma_dev2->dev, tr->addr[i])) {
+			dev_warn(dev, "dma_map_page() 4 failed\n");
+			pmem_dma_unmap(tr);
+			return false;
+		}
+		i++;
+	}
+
+	/* Issue the local PMEM I/O. */
+	i = 0;
+	flags = DMA_CTRL_ACK;
+	cb = NULL;
+	bio_for_each_bvec(bvec, bio, iter) {
+		if (iter.bi_size == bvec.bv_len) {
+			flags |= DMA_PREP_INTERRUPT;
+			cb = pmem_dma_callback;
+		}
+		if (op_is_write(bio_op(bio))) {
+			tx = dmaengine_prep_dma_memcpy(dma_chan1,
+			    tr->laddr + (iter.bi_sector - sector) * 512,
+			    tr->addr[i], bvec.bv_len, flags);
+		} else {
+			tx = dmaengine_prep_dma_memcpy(dma_chan1, tr->addr[i],
+			    tr->laddr + (iter.bi_sector - sector) * 512,
+			    bvec.bv_len, flags);
+		}
+		if (!tx) {
+			dev_warn(dev, "dmaengine_prep_dma_memcpy() 1 failed\n");
+			goto error;
+		}
+		tx->callback_result = cb;
+		tx->callback_param = tr;
+		cookie = dmaengine_submit(tx);
+		if (dma_submit_error(cookie)) {
+			dev_warn(dev, "dmaengine_submit() 1 failed\n");
+			goto error;
+		}
+		last_cookie1 = cookie;
+		i += dmas;
+	}
+	dma_async_issue_pending(dma_chan1);
+
+	if (!raddr)
+		return true;
+
+	/* Issue the remote PMEM I/O. */
+	i = 1;
+	flags = DMA_CTRL_ACK;
+	cb = NULL;
+	bio_for_each_bvec(bvec, bio, iter) {
+		if (iter.bi_size == bvec.bv_len) {
+			flags |= DMA_PREP_INTERRUPT;
+			cb = pmem_dma_callback;
+		}
+		tx = dmaengine_prep_dma_memcpy(dma_chan2,
+		    tr->raddr + (iter.bi_sector - sector) * 512,
+		    tr->addr[i], bvec.bv_len, flags);
+		if (!tx) {
+			dev_warn(dev, "dmaengine_prep_dma_memcpy() 2 failed\n");
+			goto error2;
+		}
+		tx->callback_result = cb;
+		tx->callback_param = tr;
+		cookie = dmaengine_submit(tx);
+		if (dma_submit_error(cookie)) {
+			dev_warn(dev, "dmaengine_submit() 2 failed\n");
+			goto error2;
+		}
+		last_cookie2 = cookie;
+		i += 2;
+	}
+	dma_async_issue_pending(dma_chan2);
+	return true;
+
+error:
+	/*
+	 * Some error has happened during the local PMEM I/O issue.
+	 * Since none of issued transactions had callback, just wait for
+	 * them to complete, free memory and fall back to software copy.
+	 */
+	if (last_cookie1) {
+		dma_async_issue_pending(dma_chan1);
+		dma_sync_wait(dma_chan1, last_cookie1);
+	}
+	pmem_dma_unmap(tr);
+	return false;
+
+error2:
+	/*
+	 * Some error has happened during the remote PMEM I/O issue.
+	 * The local PMEM I/O is already running and we can't stop it.
+	 * Since none of issued remote transactions had callback, wait for
+	 * them to complete and return I/O error when local I/O completes.
+	 */
+	if (last_cookie2) {
+		dma_async_issue_pending(dma_chan2);
+		dma_sync_wait(dma_chan2, last_cookie2);
+	}
+	pmem_dma_callback(tr, &dummy_result);
+	return true;
+}
+
 static int pmem_open(struct block_device *bdev, fmode_t mode)
 {
 	struct pmem_device *pmem = bdev->bd_disk->private_data;
 	struct pmem_label *label = pmem->label;
 
+	if ((pmem->opened++) == 0)
+		pmem_dma_init(pmem);
```

Comment:
Do we want to report an error from ```pmem_open()``` if ```pmem_dma_init()``` failed to allocate a channel?

Reply:
If you mean console message, then pmem_dma_init() can potentially report it in more details.  If you mean pmem_open() should return error -- then no, code should just fall back to software copy.  It should not happen on our systems, but but I left the console message for now to ease the testing. I may remove it sometimes later, since I know few people do use TrueNAS with NVDIMMs in custom systems, that may theoretically not have DMA engine.